### PR TITLE
cleanup hyparview

### PR DESCRIPTION
**Pull request by mschuwalow**
- [x] Migrate remaining parts of activeProtocol
- [x] update views to expose a single send function that will either enqueue to an active connection or open a new one
   ```
   def send(node: NodeAddress, message: Message): STM[Nothing, Unit]
   def newMessages: Stream[Nothing, (NodeAddress, Message)]
    ```
- [x] Harden initial protocol
- [ ] Harden active protocol
- [ ] Harden plumtree
- [x] Expand on MockConnection / testing functionality
- [x] Rip out old implementations
- [ ] ~Migrate Plumtree~

### CODE REVIEW

File: keeper/src/main/scala/zio/keeper/hyparview/protocols.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+              remoteAddress <- remoteAddressRef.get.commit
+              _ <- remoteAddress.fold[ZIO[R, Nothing, Unit]](ZIO.unit) { remoteAddress =>
+                    val complete =
+                      ZSTM.ifM(disconnected.succeed(()))(
+                        Views.removeFromActiveView(remoteAddress) *> {
+                          if (keep) Views.addToPassiveView(remoteAddress) else STM.unit
+                        }.as(true),
+                        STM.succeed(false)
+                      )
+                    complete.commit.flatMap {
```
Comment (mschuwalow) [inline keeper/src/main/scala/zio/keeper/hyparview/protocols.scala:96]: when

File: keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala [hunk 2]
```diff
@@ -17,19 +20,35 @@ object HyParViewConfig {
...
+         |shuffleTTL: $shuffleTTL
+         |connectionBuffer: $connectionBuffer
+         |userMessagesBuffer: $userMessagesBuffer
+         |concurrentIncomingConnections: $concurrentIncomingConnections""".stripMargin
   }
 
+  trait Service {
+    val getConfigSTM: STM[Nothing, Config]
+
+    val getConfig: UIO[Config] =
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala:40]: Does it make sense to expose this?

File: keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala [hunk 3]
```diff
@@ -17,19 +20,35 @@ object HyParViewConfig {
...
+         |shuffleNActive: $shuffleNActive
+         |shuffleNPassive: $shuffleNPassive
+         |shuffleTTL: $shuffleTTL
+         |connectionBuffer: $connectionBuffer
+         |userMessagesBuffer: $userMessagesBuffer
+         |concurrentIncomingConnections: $concurrentIncomingConnections""".stripMargin
   }
 
+  trait Service {
+    val getConfigSTM: STM[Nothing, Config]
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala:38]: Use `USTM`, less typing.

File: keeper/src/main/scala/zio/keeper/hyparview/PeerService.scala [hunk 4]
```diff
@@ -1,37 +1,70 @@
...
+import zio.logging.Logging
 
 object PeerService {
 
+  trait Service {
+    val identity: UIO[NodeAddress]
+    val getPeers: UIO[Set[NodeAddress]]
+    def send(to: NodeAddress, message: PeerMessage): IO[Nothing, Unit]
+    val receive: Stream[Nothing, (NodeAddress, PeerMessage)]
+    val events: Stream[Nothing, PeerEvent]
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/hyparview/PeerService.scala:20]: Does it make sense to unify these?

File: keeper/src/main/scala/zio/keeper/hyparview/Views.scala [hunk 5]
```diff
@@ -1,251 +1,217 @@
...
-          passiveViewCapacity <- passiveViewCapacity
-          _ <- if (node == myself0 || inActive || inPassive) STM.unit
-              else {
+
+        override def addToActiveView(
+          node: NodeAddress,
+          send: Message => STM[Nothing, _],
+          disconnect: STM[Nothing, _]
+        ): STM[Nothing, Unit] =
+          if (node == myself0) STM.unit
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/hyparview/Views.scala:124]: Moar "whens" :).

File: keeper/src/main/scala/zio/keeper/transport/Connection.scala [hunk 6]
```diff
@@ -1,10 +1,197 @@
...
+            val receive: ZStream[R, E, O] =
+              self.receive
+
+            val close: UIO[Unit] =
+              self.close
+          }
+        }
+    }
+
+  def biMap[I1, O1](f: I1 => I, g: O => O1): Connection[R, E, I1, O1] =
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/transport/Connection.scala:57]: Nah, just lowercase `M`.

File: keeper/src/main/scala/zio/keeper/transport/Connection.scala [hunk 7]
```diff
@@ -1,10 +1,197 @@
...
+    new Connection[R, E, I1, O1] {
+      def send(data: I1): ZIO[R, E, Unit] = self.send(f(data))
+
+      val receive: ZStream[R, E, O1] = self.receive.map(g)
+
+      val close: UIO[Unit] = self.close
```
Comment (mijicd) [inline keeper/src/main/scala/zio/keeper/transport/Connection.scala:63]: Does it make sense to have `close` here, or should it be pushed to some managed scope outside?

**Review by mijicd — COMMENTED** [review]
Just a very high first pass, I need more time to dig into this one.

File: keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala (hunk 2 above)
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala:40]: Removed. Just the function in the object is enough

File: keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala (hunk 3 above)
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/hyparview/HyParViewConfig.scala:38]: done, thanks 👍

File: keeper/src/main/scala/zio/keeper/hyparview/PeerService.scala (hunk 4 above)
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/hyparview/PeerService.scala:20]: yes, will do

File: keeper/src/main/scala/zio/keeper/transport/Connection.scala (hunk 7 above)
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/transport/Connection.scala:63]: Close will be removed

Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/transport/Connection.scala]: close has been removed

**Comment by mschuwalow** [comment]
There is still work to be done here, but I believe it's fine as follow up PRs.

**Review by mijicd — APPROVED** [review]
Just a few minor comments, good to go otherwise.

File: examples/src/main/scala/zio/keeper/example/K8sExample.scala [hunk 8]
```diff
@@ -21,7 +21,7 @@ object K8sTestNode extends zio.App {
   val discovery =
     ZLayer.fromManaged(
       for {
-        config <- ZManaged.environment[Config[SwimConfig]].map(_.get)
+        config <- ZManaged.environment[ZConfig[SwimConfig]].map(_.get)
```
[the remark this answers is not in the document]
Comment (mschuwalow) [reply examples/src/main/scala/zio/keeper/example/K8sExample.scala]: Zio config exposes `getConfig`. That's even nicer

File: keeper/src/main/scala/zio/keeper/hyparview/Views.scala [hunk 9]
```diff
@@ -1,251 +1,198 @@
...
-    def myself: STM[Nothing, NodeAddress]
-    def activeViewCapacity: STM[Nothing, Int]
-    def passiveViewCapacity: STM[Nothing, Int]
-
-    def activeView: STM[Nothing, Set[NodeAddress]]
-    def passiveView: STM[Nothing, Set[NodeAddress]]
-
-    final def activeViewSize: STM[Nothing, Int] =
-      activeView.map(_.size)
+    def activeView: USTM[Set[NodeAddress]]
```
[the remark this answers is not in the document]
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/hyparview/Views.scala:12]: I want activeView and passiveView to have the same api, but activeView is represented by a TMap internally and I don't think there is a nice way to link these two.
Plus I think we only allow modification through the service here

File: keeper/src/main/scala/zio/keeper/hyparview/Views.scala [hunk 10]
```diff
@@ -1,251 +1,198 @@
...
-      def addToPassiveView(node: NodeAddress): STM[Nothing, Unit] =
-        for {
-          inActive            <- activeView0.contains(node)
-          inPassive           <- passiveView0.contains(node)
-          passiveViewCapacity <- passiveViewCapacity
-          _ <- if (node == myself0 || inActive || inPassive) STM.unit
-              else {
+
+        override def addToActiveView(node: NodeAddress, send: Message => USTM[_], disconnect: USTM[_]): USTM[Unit] =
+          STM.when(!(node == config.address)) {
```
[the remark this answers is not in the document]
Comment (mschuwalow) [reply keeper/src/main/scala/zio/keeper/hyparview/Views.scala]: done

File: examples/src/main/scala/zio/keeper/example/K8sExample.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (mijicd) [reply examples/src/main/scala/zio/keeper/example/K8sExample.scala]: True.

**Pull request merged**